Reflow Documentation V1 (#3970)
sqlfluff/sqlfluff: docs/source/layout.rst

it's going to be **whitespace** (closely followed by **cApiTaLiSaTiOn** 😁).

More specifically, **whitespace** divides into three key themes:

#. **Spacing**: The amount of whitespace between elements on the same line.

#. **Line Breaks**: The choice of where within the code it is inappropriate,
   appropriate or even compulsory to have a line break.

#. **Indentation**: Given a line break, how much whitespace should precede
   the first code element on that line.

*SQLFluff* aims to be *opinionated* on this theme, but also *configurable*
(see :ref:`layoutconfig`). The tool will have a default viewpoint and will aim
to have views on all of the important aspects of SQL layout, but if you
(or your organisation) don't like those views then we aim to allow enough
configuration that you can lint in line with your views, and still use
*SQLFluff*. For more information on how to configure rules to your own
viewpoint see :ref:`config`.

.. note::

    This section of the docs handles the intent and reasoning behind how
    layout is handled by SQLFluff. For a deeper look at how this is achieved
    internally see :ref:`reflowinternals`.

Spacing
-------

> Of the different elements of whitespace, spacing it likely the least

Reviewer: ```suggestion
Of the different elements of whitespace, spacing is likely the least
```

> controversial. By default all elements are separated by a single space

Reviewer: ```suggestion
controversial. By default, all elements are separated by a single space
```

character. Except for very specific circumstances (see section on
:ref:`alignedelements`), any additional space between elements is
usually unwanted and a distraction for the reader. There are however
several common cases where *no whitespace* is more appropriate, which
fall into two cases.

#. *No whitespace but a newline is allowed.* This option is configured

>    using the :code:`touch` setting. The most common example of

Author: yeah good point

   this is the spacing around commas. For example :code:`SELECT a , b`
   would be unusual and more normally be written :code:`SELECT a, b`.
   Inserting a newline between the :code:`a` and comma would not
   cause issues and may even be desired, for example:

   .. code-block:: sql

      SELECT
         col_a
         , col_b
         -- Newline present before column
         , col_c
         -- When inline, comma should still touch element before.
         , GREATEST(col_d, col_e) as col_f
      FROM tbl_a

#. *No whitespace and a newline is not allowed.* This option is

>    configured using the :code:`inline` setting. The most common example

Reviewer: Same question as above.

   of this is spacing within the parts of qualified identifier e.g.
   :code:`my_schema.my_table`. If a newline were present between the
   :code:`.` and either :code:`my_schema` or :code:`my_table`, then
   the expression would not parse and so no newlines should be allowed.

.. _alignedelements:

Aligned elements
^^^^^^^^^^^^^^^^

A special case of spacing is where elements are set to be aligned
within some limits. This is not enabled by default, but can be
be configured to achieve layouts like:

.. code-block:: sql

   SELECT
      a           AS first_column,
      b           AS second_column,
      (a + b) / 2 AS third_column
   FROM foo AS bar

In this example, the alias expressions are all aligned with each other.

> Practically to configure this, SQLFluff needs to know what elements to

Reviewer: Is the word "practically" needed here?

Author: not really - I'll remove it.

align and how far to search to find elements which should be aligned
with eachother. The configuration to achieve this layout is:

.. code-block:: ini

   [sqlfluff:layout:type:alias_expression]
   # We want non-default spacing _before_ the alias expressions.
   spacing_before = align
   # We want to align them within the next outer select clause.
   # This means for example that alias expressions within the FROM
   # or JOIN clause would _not_ be aligned with them.

>    align_within = select_clause

Reviewer: Wondering if `align_scope` might be clearer.

Author: I actually think `align_scope` is a better name for `align_boundary`. I've changed it - do you think it's a bit clearer now?

Reviewer: Looks good!

   # The point at which to stop searching outward for siblings, which
   # in this example would likely be the boundary of a CTE. Stopping
   # when we hit brackets is usually a good rule of thumb for this
   # configuration.

>    align_boundary = bracketed

Author: I'll add an example

>    align_boundary = bracketed

Author: I've added a relatively long example here, let me know what you think 👍

Line Breaks
-----------

When controlling line breaks we are trying to achieve a few different things:

#. Do we have *enough* line breaks that *line length* doesn't become
   excessive. Long lines are hard to read, especially given that readers
   may be on varying screen sizes or have multiple windows open. This is
   (of course) configurable, but the default is 80 characters (in line with
   the `dbt Labs SQL style guide`_.)

> #. Are the positioning of *blank lines* (i.e. lines with nothing other

Reviewer: ```suggestion
#. Is the positioning of *blank lines* (i.e. lines with nothing other
```

   than whitespace on them) appropriate. There are some circumstances
   where a blank line is *desired* (e.g. between CTEs). There are others
   where they are not, in particular *multiple blank lines*, for example
   at the beginning of a file.

#. Where we do have line breaks, are they positioned appropriately and
   consistently with regards to other elements around them. This is most
   common when it comes to *commas*, and whether they should be *leading*
   (e.g. :code:`, my_column`) or *trailing* (e.g. :code:`my_column,`). In
   less common cases it may also be desirable for some elements to have both

>    and line break *before and after* (e.g. a set operator such as `UNION`).

Reviewer: The words "both and" read a bit strange to me. I think it's technically correct, but it "surprises" me when I read it.

Could we say something like:
"to have both -- i.e. a line break before and after..."

.. code-block:: sql

   SELECT
      nested_within_select AS first_column,
      some_function(
         nested_within_function,
         also_nested_within_function
      ) AS indented_the_same_as_opening_bracket
   FROM indented_the_same_as_select

Comment Indents
^^^^^^^^^^^^^^^

**Comments** are dealt with differently, depending on whether they're
*block* comments (:code:`/* like this */`), which might optionally
include newlines, or *inline* comments (:code:`-- like this`) which
are necessarily only on one line.

*  *Block comments* cannot share a line with any code elements (so
   in effect they must start on their own new line), they cannot be
   followed by any code elements on the same line (and so in effect
   must be followed by a newline, if we are to avoid trailing
   whitespace). None of the lines within the block comment may have
   an indent less than the first line of the block comment (although
   additional indentation within a comment is allowed), and that first
   line should be aligned with the first code element *following*
   the block comment.

   .. code-block:: sql

      SELECT
         /* This is a block comment starting on a new line

>          which contains a newline (continuing with at least

Author: This isn't implemented yet - I think it should be, that would be nice. I'll add a note to the docs to say it's not done yet. I might be able to pick this up in the reflow work.

first nested element, but subsequent elements are indented to
match the line position of that first element. Two examples might be:

.. code-block:: sql

   -- A select statement with two hanging indents:
   SELECT no_line_break_before_me,
          indented_to_match_the_first,
          1 + (a
               + b) AS another_more_complex_example
   FROM my_table;

   -- This TSQL example is also in essence a hanging indent:
   DECLARE @prv_qtr_1st_dt DATETIME,
           @last_qtr INT,
           @last_qtr_first_mn INT,
           @last_qtr_yr INT;

In some circumstances this layout can be quite neat (the
:code:`DECLARE` statement is a good example of this), however
once indents are nested or indentation styles are mixed it
can rapidly become confusing (as partially shown in the first
example). Additionally, unless the leading element of the first
line is very short, hanging indents use much *larger indents*
than a traditional simple indent where a line break is used before
the first element.

Hanging indents have been supported in SQLFluff up to the 1.x
versions, however **they will no longer by supported from 2.0.0**
onwards. This is due to the ambiguity which they bring to

> fixing poorly formatted sql. Take the following code:

Reviewer 2: sql -> SQL

second :code:`if` statement. If treated as a simple case, these tags
would have different indents, because they are at different levels of
the parse tree and so clearly there is a conflict to be resolved.

The view SQLFluff takes on how to resolve this conflict is to pull
all of the tags in this section down to the indent of the
*least indented* (in the example below that would be the closing
:code:`endif` tag). This is similar to the treatment of
`C Preprocessor Directives`_, which are treated somewhat as being
outside the structure of the rest of the file. In these cases,
the content is also *not further indented* as in the simple case
because it makes it harder to line up elements within the affected
section and outside (in the example below the :code:`SELECT` and
:code:`FROM` are a good illustration).

.. code-block:: SQL+Jinja

   SELECT
      a,
      {% if True %}
         -- This is a simple case. The opening and closing tag are
         -- both at the same level within the SELECT clause.
         simple_case AS example,
      {% endif %}
      b,
   {% if True %}
      -- This is a complex case. The opening tag is within the SELECT

>       -- clause, but the closing tag is outside the statement

Reviewer: 😎 💯

Reviewer 3: Real nice

These can also be combined, so if :code:`indented_using_on` config is set to
:code:`False`, and :code:`indented_on_contents` is also set to :code:`False`
then the SQL would become:

.. code-block:: sql

   SELECT
      a,
      b
   FROM my_table
   JOIN another_table
   ON condition1
   AND condition2

There is also a similar :code:`indented_ctes` config (defaulted to
:code:`False`) which can be set to :code:`True` to enforce CTEs to be
indented within the :code:`WITH` clause:

.. code-block:: sql

   WITH
      some_cte AS (
         SELECT 1 FROM table1
      ),

      some_other_cte AS (
         SELECT 1 FROM table1
      )

   SELECT 1 FROM some_cte

By default, *SQLFluff* aims to follow the indentation most common approach
to indentation. However, if you have other versions of indentation which are
supported by published style guides, then please submit an issue on GitHub
to have that variation supported by *SQLFluff*.

.. _layoutspacingconfig:

Configuring layout and spacing
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :code:`[sqlfluff:layout]` section of the config controls the treatment of
spacing and line breaks across all rules. The syntax of this section is very

> expressive, however in normal use, only very small alterations should be

Reviewer: ```suggestion
expressive; however in normal use, only very small alterations should be
```

The syntax of the section headings here select by *type*, which corresponds
to the :code:`type` defined in the dialect. For example the following section
applies to elements of the *type* :code:`comma`, i.e. :code:`,`.

.. code-block:: cfg

   [sqlfluff:layout:type:comma]
   spacing_before = touch
   line_position = trailing

Within these configurable sections there are a few key elements which are
available:

*  **Spacing Elements**: :code:`spacing_before`, :code:`spacing_after` and
   :code:`spacing_within`. For each of these options there are a few possible
   settings:

   *  The default spacing for all elements is :code:`single` unless otherwise
      specified. In this state, elements will be spaced with a single space
      character unless there is a line break between them.

   *  The value of :code:`touch` allows line breaks, but if no line break is
      present, then no space should be present. A great example of this is
      the spacing before commas (as shown in the config above), where line
      breaks may be allowed, but if not they should *touch* the element before.

   *  The value of :code:`inline` is effectively the same as :code:`touch`

>       but in addition, no line breaks are allowed. This best illustrated

Reviewer: ```suggestion
      but in addition, no line breaks are allowed. This is best illustrated
```

>       by the :code:`spacing_within` a qualified identifier like

Reviewer: The wording here seems strange -- can you review? "by the XXX a qualified identifier"

      :code:`my_schema.my_table`.

*  **Line Position**: set using the :code:`line_position` option. By default
   this is unset, which implies no particular line position requirements. The
   available options are:

   *  :code:`trailing` and :code:`leading`, which are most common in the
      placement of commas. Both of these settings *also* allow the option

>       of a comma on it's own on a line, or in the middle of a line, *but*

Reviewer: ```suggestion
      of a comma on its own on a line, or in the middle of a line, *but*
```

      if there is a line break on *either side* then they make sure it's
      on the *correct side*. By default we assume *trailing* commas, but if
      you (or your organisation) have settled on *leading* commas then
      you should add the following section to your config:

      .. code-block:: cfg

         [sqlfluff:layout:type:comma]
         line_position = leading

>    *  :code:`alone`, which means if there is a line break on either side

Reviewer: ```suggestion
   *  :code:`alone`, which means if there is a line break on either side,
```

         , col_c
         -- When inline, comma should still touch element before.
         , GREATEST(col_d, col_e) as col_f
      FROM tbl_a

#. *No whitespace and a newline is not allowed.* This option is
   configured using the :code:`inline` option in the :code:`spacing_*`
   configuration settings. The most common example of this is spacing
   within the parts of qualified identifier e.g. :code:`my_schema.my_table`.
   If a newline were present between the :code:`.` and either
   :code:`my_schema` or :code:`my_table`, then the expression would not
   parse and so no newlines should be allowed.

.. _alignedelements:

Aligned elements
^^^^^^^^^^^^^^^^

A special case of spacing is where elements are set to be aligned
within some limits. This is not enabled by default, but can be
be configured to achieve layouts like:

.. code-block:: sql

   SELECT
      a           AS first_column,
      b           AS second_column,
      (a + b) / 2 AS third_column
   FROM foo AS bar

In this example, the alias expressions are all aligned with each other.
To configure this, SQLFluff needs to know what elements to
align and how far to search to find elements which should be aligned

> with eachother. The configuration to achieve this layout is:

Reviewer 3: ```suggestion
with each other. The configuration to achieve this layout is:
```

   # This means for example that alias expressions within the FROM
   # or JOIN clause would _not_ be aligned with them.
   align_within = select_clause
   # The point at which to stop searching outward for siblings, which
   # in this example would likely be the boundary of a CTE. Stopping
   # when we hit brackets is usually a good rule of thumb for this
   # configuration.
   align_scope = bracketed

Of these configuration values, the :code:`align_scope` is potentially
the least obvious. The following example illustrates the impact it has.

.. code-block:: sql

   -- With
   --    align_scope = bracketed
   --    align_within = select_clause

   WITH foo as (
      SELECT
         a,
         b,
         c     AS first_column
         d + e AS second_column
   )

   SELECT
      a           AS first_column,
      (a + b) / 2 AS third_column
   FROM foo AS bar;

   -- With
   --    align_scope = bracketed
   --    align_within = statement

   WITH foo as (
      SELECT
         a,
         b,
         c     AS first_column
         d + e AS second_column
   )

   SELECT
      a           AS first_column,
      (a + b) / 2 AS third_column

>    FROM foo       AS bar

Author: good idea - I might have to use a valid sql comment format otherwise pygments will complain but it's a good idea 👍

once indents are nested or indentation styles are mixed it
can rapidly become confusing (as partially shown in the first
example). Additionally, unless the leading element of the first
line is very short, hanging indents use much *larger indents*
than a traditional simple indent where a line break is used before
the first element.

Hanging indents have been supported in SQLFluff up to the 1.x
versions, however **they will no longer by supported from 2.0.0**
onwards. This is due to the ambiguity which they bring to
fixing poorly formatted SQL. Take the following code:

.. code-block:: sql

   SELECT   this_is,
   badly_formatted, code_and,
      not_obvious,
         what_was,
   intended FROM my_table

Given the lack of line break between :code:`SELECT` and
:code:`this_is`, it would appear that the user is intending
a hanging indent, however it is also plausible that they did
not and they just forgot to add a line break between them.
This ambiguity is unhelpful, both for SQLFluff as a tool,
but also for people who write SQL that there two ways of
indenting their SQL. Given SQLFluff aims to provide consistency
in SQL layout and remove some of the burden of needing to make

> choices like this - and that nobody is thinking of using *only*

Reviewer 3: I'm totally fine to ditch hanging indents (they don't make sense beyond basic SQL examples), but not sure I follow this:

> nobody is thinking of using *only* hanging indents

Is that because you too are implying they only work for some scenarios and therefore no one uses them consistently, making it unhelpful for SQLFluff to try and apply it universally (which is what config is for)?

Author: I'm saying that whenever they're used - it's usually in combination with traditional indents. There aren't any projects where _traditional_ indents aren't allowed.

Author: I'll make that a little clearer before merging.
